server/neptune/temporalworker/server.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/uber-go/tally/v4"
+	"github.com/urfave/cli"
+	"github.com/urfave/negroni"
+	"go.temporal.io/sdk/client"
+	temporal_tally "go.temporal.io/sdk/contrib/tally"
+	"go.temporal.io/sdk/worker"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+const (
+	AtlantisNamespace        = "atlantis"
+	DeployTaskqueue          = "deploy"
+	ProjectJobsViewRouteName = "project-jobs-detail"
+)
+
+// Config is TemporalWorker specific user config
+type Config struct {
+	AtlantisURL      *url.URL
+	AtlantisVersion  string
+	CtxLogger        logging.Logger
+	SslCertFile      string
+	SslKeyFile       string
+	TemporalHostPort string
+	Scope            tally.Scope
+	Closer           io.Closer
+	Port             int
+}
+
+type Server struct {
+	Logger           logging.Logger
+	HttpServerProxy  *neptune_http.ServerProxy
+	Port             int
+	StatsScope       tally.Scope
+	StatsCloser      io.Closer
+	TemporalHostPort string
+}
+
+// TODO: as more behavior is added into the TemporalWorker package, inject corresponding dependencies
+func NewServer(config *Config) (*Server, error) {
+	jobsController := &controllers.JobsController{
+		AtlantisVersion:     config.AtlantisVersion,
+		AtlantisURL:         config.AtlantisURL,
+		KeyGenerator:        controllers.JobIDKeyGenerator{},
+		StatsScope:          config.Scope,
+		Logger:              config.CtxLogger,
+		ProjectJobsTemplate: templates.ProjectJobsTemplate,
+	}
+	// router initialization
+	router := mux.NewRouter()
+	router.HandleFunc("/healthz", Healthz).Methods("GET")
+	router.HandleFunc("/jobs/{job-id}", jobsController.GetProjectJobs).Methods("GET").Name(ProjectJobsViewRouteName)
+	router.HandleFunc("/jobs/{job-id}/ws", jobsController.GetProjectJobsWS).Methods("GET")
+	n := negroni.New(&negroni.Recovery{
+		Logger:     log.New(os.Stdout, "", log.LstdFlags),
+		PrintStack: false,
+		StackAll:   false,
+		StackSize:  1024 * 8,
+	})
+	n.UseHandler(router)
+	httpServerProxy := &neptune_http.ServerProxy{
+		SSLCertFile: config.SslCertFile,
+		SSLKeyFile:  config.SslKeyFile,
+		Server:      &http.Server{Addr: fmt.Sprintf(":%d", config.Port), Handler: n},
+		Logger:      config.CtxLogger,
+	}
+	server := Server{
+		Logger:           config.CtxLogger,
+		HttpServerProxy:  httpServerProxy,
+		Port:             config.Port,
+		StatsScope:       config.Scope,
+		StatsCloser:      config.Closer,
+		TemporalHostPort: config.TemporalHostPort,
+	}
+	return &server, nil
+}
+
+func (s Server) Start() error {
+	defer s.Logger.Close()
+
+	// temporal client + worker initialization
+	temporalClient, err := s.buildTemporalClient()
+	if err != nil {
+		return err
+	}
+	defer temporalClient.Close()
+	go func() {
+		w := worker.New(temporalClient, DeployTaskqueue, worker.Options{
+			// ensures that sessions are preserved on the same worker
+			EnableSessionWorker: true,
+		})
+		if err := w.Run(worker.InterruptCh()); err != nil {
+			log.Fatalln("unable to start worker", err)
+		}
+	}()
```

Comment:
Do you know how the interrupt part works here? like our server listens to sigterms but is that what the worker is doing with `worker.InterruptCh`? Also if these two shutdown independently, shouldn't we have a wait group in the mainloop to make sure we're gracefully shutting down?

Reply:
I only just realized this after looking at it.

Reply:
also if our worker fails to start i think we should shutdown the server, but probably not the other way around.

Reply:
According to the SDK: `the function Run(interruptCh <-chan interface{}) error` will run the worker in a blocking fashion and stop the worker when interruptCh receives signal. Specifically, worker.InterruptCh() will stop the worker given a SIGINT or SIGTERM, which matches what we are looking for. The waitgroup is a good idea though, I'll add it.

Reply:
re: `also if our worker fails to start i think we should shutdown the server, but probably not the other way around.` Currently, we fail and exit if the worker doesn't successfully start (i.e. `Run()` returns err). Also, we only log an error if our server fails to `ListenAndServe()`.